code/modules/power/chem_combustion.dm

```diff
@@ -0,0 +1,394 @@
+
+		src.updateUsrDialog()
+		return
+
+#define FUEL_QUALITY_VERY_HIGH 10
+#define FUEL_QUALITY_HIGH 6
+#define FUEL_QUALITY_AVERAGE 3
+#define FUEL_QUALITY_LOW 2
+#define FUEL_QUALITY_VERY_LOW 1
+
+	process()
+		if (!src || !src.active)
+			return
+
+		var/average_volatility = get_average_volatility(src.fuel_tank.reagents)
+		var/available_oxygen = src.check_available_oxygen()
+		if (!available_oxygen || !average_volatility)
+			src.stop_engine()
+			src.visible_message("<span class='alert'>The [src]'s engine fails to run, it has nothing to combust!</span>")
+			return
+
+		if (!src.anchored)
+			src.stop_engine()
+			src.visible_message("<span class='alert'>The [src] makes a horrible racket and shuts down, it has become unanchored!</span>")
+			return
+
+		var/obj/cable/C = src.get_output_cable()
+		if (!C)
+			src.stop_engine()
+			src.visible_message("<span class='alert'>Electricity begins to arc off the [src] causing it to shutdown, it has nothing to output to!</span>")
+			elecflash(src.loc, 0, power = 3, exclude_center = 0)
+			return
+
+		src.last_output = ((average_volatility * (available_oxygen * 5)) * src.output_multiplier) KILO WATTS
+		var/datum/powernet/P = C.get_powernet()
+		P.newavail += src.last_output
+
+		var/turf/simulated/T = get_turf(src)
+		if (istype(T))
+			var/datum/gas_mixture/payload = new /datum/gas_mixture
+			payload.carbon_dioxide = src.atmos_drain_rate * average_volatility * src.output_multiplier
+			payload.temperature = T20C
+			T.assume_air(payload)
+
+		if (src.inlet_tank)
+			src.inlet_tank.air_contents.remove(src.atmos_drain_rate * average_volatility * src.output_multiplier)
+		else
+			T.air.remove(src.atmos_drain_rate * average_volatility * src.output_multiplier)
+
+		src.fuel_tank.reagents.remove_any(src.fuel_drain_rate * src.output_multiplier)
+
+		src.UpdateIcon()
+		src.updateDialog()
+
+	proc/get_average_volatility(datum/reagents/R)
+		if (!R || !R.total_volume)
+			return
+
+		var/average = 0
+		for (var/reagent_id in R.reagent_list)
+			var/datum/reagent/current_reagent = R.reagent_list[reagent_id]
+			if (current_reagent)
+				switch(reagent_id)
+					if ("foof", "kerosene", "nitrotri_dry", "dbreath")
+						average += (FUEL_QUALITY_VERY_HIGH * current_reagent.volume)
+
+					if ("blackpowder", "phlogiston", "napalm_goo", "sorium", "firedust")
+						average += (FUEL_QUALITY_HIGH * current_reagent.volume)
+
+					if ("fuel", "acetone")
+						average += (FUEL_QUALITY_AVERAGE * current_reagent.volume)
+
+					if ("oil", "butter", "diethylamine", "ethanol")
+						average += (FUEL_QUALITY_LOW * current_reagent.volume)
+
+					if ("plasma", "magnesium", "phosphorus")
+						average += (FUEL_QUALITY_VERY_LOW * current_reagent.volume)
+
+					else
+						var/datum/reagent/fooddrink/alcoholic/current_alcoholic_reagent = current_reagent
+						if (istype(current_alcoholic_reagent) && current_alcoholic_reagent.alch_strength >= 0.5)
+							average += 3
```

Comment:
List/Table lookups would be "way faster and open it up for more finely tuned values.

Reply:
Done